libyul/backends/wasm/WordSizeTransform.cpp

```diff
@@ -0,0 +1,238 @@
+								varDecl.location,
+								{TypedName{varDecl.location, newLhs[i], "u64"_yulstring}},
+								std::move(newRhs[i])
+							}
+						);
+					return ret;
+				}
+				else
+					yulAssert(false, "");
+			}
+			else if (_s.type() == typeid(Assignment))
+			{
+				Assignment& assignment = boost::get<Assignment>(_s);
+				yulAssert(assignment.value, "");
+				if (
+					assignment.value->type() == typeid(FunctionalInstruction) ||
+					assignment.value->type() == typeid(FunctionCall)
+				)
+				{
+					if (assignment.value) visit(*assignment.value);
+					rewriteIdentifierList(assignment.variableNames);
+					return boost::none;
+				}
+				else if (
+					assignment.value->type() == typeid(Identifier) ||
+					assignment.value->type() == typeid(Literal)
+				)
+				{
+					yulAssert(assignment.variableNames.size() == 1, "");
+					auto newRhs = expandValue(*assignment.value);
+					YulString lhsName = assignment.variableNames[0].name;
+					vector<Statement> ret;
+					for (int i = 0; i < 4; i++)
+						ret.push_back(
+							Assignment{
+								assignment.location,
+								{Identifier{assignment.location, m_variableMapping.at(lhsName)[i]}},
+								std::move(newRhs[i])
+							}
+						);
+					return ret;
+				}
+				else
+					yulAssert(false, "");
+			}
+			else
+				visit(_s);
+			return boost::none;
+		}
+	);
+}
+
+array<YulString, 4> WordSizeTransform::generateU64IdentifierNames(YulString const& _s)
+{
+	yulAssert(m_variableMapping.find(_s) == m_variableMapping.end(), "");
+	array<YulString, 4> ret;
+	for (int i = 0; i < 4; i++)
+	{
+		auto newName = m_nameDispenser.newName(_s);
+		m_variableMapping[_s][i] = newName;
+		ret[i] = newName;
+	}
+	return ret;
+}
+
+array<unique_ptr<Expression>, 4> WordSizeTransform::expandValue(Expression const& _e)
+{
+	array<unique_ptr<Expression>, 4> ret;
+	if (_e.type() == typeid(Identifier))
+	{
+		Identifier const& id = boost::get<Identifier>(_e);
+		for (int i = 0; i < 4; i++)
+			ret[i] = make_unique<Expression>(Identifier{id.location, m_variableMapping.at(id.name)[i]});
+	}
+	else if (_e.type() == typeid(Literal))
+	{
+		Literal const& lit = boost::get<Literal>(_e);
+		yulAssert(lit.kind == LiteralKind::Number, "Non-numeric literal unimplemented.");
+		u256 val = u256(lit.value.str());
+		for (int i = 3; i >= 0; i--)
+		{
+			u256 currentVal = val & std::numeric_limits<uint64_t>::max();
+			val >>= 64;
+			ret[i] = make_unique<Expression>(
+				Literal{
+					lit.location,
+					LiteralKind::Number,
+					YulString(currentVal.str()),
+					"u64"_yulstring
+				}
+			);
+		}
+	}
+	else
+		yulAssert(false, "");
+	return ret;
+}
+
+void WordSizeTransform::rewriteFunctionCallArguments(vector<Expression>& _args)
+{
+	iterateReplacing(
+		_args,
+		[&](Expression& _e) -> boost::optional<vector<Expression>>
+		{
+			// ExpressionSplitter guarantees arguments to be Identifier or Literal
+			yulAssert(_e.type() == typeid(Identifier) || _e.type() == typeid(Literal), "");
+			vector<Expression> ret;
```

Comment:
Does `vector<Expression>(expandValue(_e))` work? Or maybe at least `ret += expandValue(_e)`?

Reply:
`expandValue(_e)` returns array of `unique_ptr<Expression>`. But here we want `vector<Expression>`, so I need to dereference then `push_back`.

I'm not sure why these 3 structs take `Expression` instead of `unique_ptr<Expression>`.
https://github.com/ethereum/solidity/blob/e9f41d1148d0442d32f0b0553b9cf742a23ede46/libyul/AsmData.h#L63-L67

Reply:
Ah right, I see!

`unique_ptr` is only used in case there is no other indirection like the one already provided by vector.